sdk/python/kfp/compiler/pipeline_spec_builder.py

```diff
@@ -239,70 +327,18 @@ def build_task_spec_for_task(
                         component_input_parameter)
 
         elif isinstance(input_value, str):
-            # Handle extra input due to string concat
```

Comment:
IIRC, this chunk of code is only applicable for string typed inputs, why merging the code and expand it to other input types? Also it's a bit hard to read the diff between the deleted code and the extracted. Can you try make the changes in place without refactoring, and see if it's actually necessary to expand the logic to non-string typed inputs?

Reply:
I found that this block of code could be reused for handling PipelineChannels inside of `worker_pool_specs` in addition to handling string typed inputs. Instead of copying the ~50 lines of code, I thought it'd be better to refactor the logic as a separate function `def replace_and_inject_placeholders()`. 

So I could un-refactor and duplicate the logic; I do have a slight preference for this refactoring but can go either way. wdyt?

Reply:
This name of the extracted method isn't accurate--the code does more than placeholder manipulation but also component input expansion.
The branch logic now reads like this:
```python
if isinstance(input_value, str):
    # shared code
    pipeline_task_spec.inputs.parameters[
                input_name].runtime_value.constant.string_value = input_value
elif isinstance(input_value, (int, float, bool, dict, list)):
    if isinstance(input_value, (dict, list):
          # shared code
    pipeline_task_spec.inputs.parameters[
                input_name].runtime_value.constant.CopyFrom(
                    to_protobuf_value(input_value))
else:
     raise
```
You can achieve the same goal, and even more code reuse, without extracting a shared method by:
```python
if not isinstance(input_value, (str, dict, list, int, float, bool)):
    raise

if isinstance(input_value, (str, dict, list)):
    # shared code

pipeline_task_spec.inputs.parameters[
            input_name].runtime_value.constant.CopyFrom(
                to_protobuf_value(input_value))
```

Reply:
Aside from the refactoring part, I wonder what's the case for dict and list? In case `CustomTrainingJobOp` is used, what's the `input_value` here?

Reply:
If `CustomTrainingJobOp` is used, then we pass in `worker_pool_spec` into `input_value`.

`input_value = [{'container_spec': {'image_uri': 'gcr.io/ml-pipeline/google-cloud-pipeline-components:2.5.0', 'command': ['echo'], 'args': ['foo']}, 'machine_spec': {'machine_type': {{channel:task=machine-type;name=Output;type=String;}}, 'accelerator_type': {{channel:task=accelerator-type;name=Output;type=String;}}, 'accelerator_count': {{channel:task=accelerator-count;name=Output;type=Integer;}}}, 'replica_count': 1}]`

Reply:
Thanks for the explanation. So `input_value` would be of type `list` in this case. Including `dict` in the same code path is just for future use cases not entirely necessary at this moment, right? I'm fine to include `dict` now. 

Have you seen my other comment on refactoring above? https://github.com/kubeflow/pipelines/pull/10883/files#r1636822260